### PR TITLE
WIP add sqlite state store (pending for gRPC component)

### PR DESCRIPTION
**Pull request by lotuc**
# Description

Implemented a sqlite state store.

The implementation is largely a copy of the oracle state store. And since it uses [github.com/mattn/go-sqlite3](github.com/mattn/go-sqlite3) as sqlite driver, `CGO` needs to enabled.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by lotuc** [comment]
Some lint problem related to this PR is found. Please ignore my latest push. I'll fix it soon.

File: state/sqlite/sqlitesetvalue.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	requestValue := req.Value
+	byteArray, isBinary := req.Value.([]uint8)
+	if isBinary {
+		requestValue = base64.StdEncoding.EncodeToString(byteArray)
+	}
+
+	// Convert to json string.
+	bt, _ := utils.Marshal(requestValue, json.Marshal)
```
Comment (daixiang0) [inline state/sqlite/sqlitesetvalue.go:56]: Please handle error.

File: state/sqlite/sqlitesetvalue.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// Returns nil or non-negative value, nil means never expire.
+func parseTTL(requestMetadata map[string]string, logger logger.Logger) (*int, error) {
+	if val, found := requestMetadata[metadataTTLKey]; found && val != "" {
+		parsedVal, err := strconv.ParseInt(val, 10, 0)
+		if err != nil {
+			return nil, fmt.Errorf("error in parsing ttl metadata : %w", err)
+		}
+		parsedInt := int(parsedVal)
```
Comment (daixiang0) [inline state/sqlite/sqlitesetvalue.go:144]: Still need to convert after `strconv.ParseInt`?

**Comment by lotuc** [comment]
@daixiang0 Hi, thanks for reviewing. I just fixed some problems you mentioned.

And `make lint | grep -i sqlite` now returns nothing and `make test` is also fine. 

But `golangci-lint run --max-issues-per-linter 2000 --max-same-issues 2000 | grep -i sqlite` returns some lint errors like 

```
...
state/sqlite/sqlite_integration_test.go:793:6            thelper           test helper function should start from t.Helper()
...
state/sqlite/sqlitesetvalue.go:76:2                      varnamelen        variable name 'tx' is too short for the scope of its usage
```

Should I fix them all? Some of them seems harmless like `tx` var name for a `sql.Tx`.

ps. I run the lint command in the dev container.

**Comment by lotuc** [comment]
And also, I'm not rebasing when fixing subsequent problems. Will do some rebasing & cleanup when it's finally accepted, is that ok?

**Comment by lotuc** [comment]
@daixiang0 Hi, I just passed conformation test locally in the dev container.
```
CGO_ENABLED=1 dlv test --build-flags '-v -tags=conftests' ./tests/conformance -- -test.run "TestStateConformance/sqlite"

make CGO=1 test
```

And also, `make lint | grep -i sqlite` returns nothing yet. Would you mind give it a review? Thanks a lot.

**Comment by berndverst** [comment]
@lotuc if CGO needs to be enabled then we cannot accept this. Our builds of Dapr disable CGO by default -- and this is intentional.

https://github.com/dapr/dapr/blob/master/Makefile#L25

FYI @yaron2 @artursouza

**Comment by artursouza** [comment]
Is there a way to have SQLite without CGO?

**Comment by yaron2** [comment]
@lotuc We really appreciate this PR, but enabling CGO compilation in daprd is problematic and has broader, less desirable implications on the project.

We are working on enabling gRPC enabled components for 1.8, which would allow this component to run as it's own CGO process without affecting the daprd process.

So I see the following options:

1. Change this implementation to not depend on CGO. If this is doable, we can merge to daprd right away.
2. Keep this PR open as-is and wait until gRPC components are enabled.

**Comment by lotuc** [comment]
@yaron2 Ok, I'll keep this open until gRPC components is ready.

**Comment by ItalyPaleAle** [comment]
@lotuc I've taken your code, reviewed it, and implemented it in a pluggable component that should work with Dapr 1.9! Take a look here: https://github.com/ItalyPaleAle/dapr-sqlite-statestore

I'm asking for that repo to be accepted into the [dapr-sandbox org](https://github.com/dapr/community/issues/231).

**Pull request closed without merging**